### ISSUE TITLE
Multi line quotes in the cover letter not working properly

### ISSUE THREAD

**Issue by prati0100** [gitgitgadget/gitgitgadget#161]
Hi,

In [this](https://github.com/gitgitgadget/git/pull/436) thread, the author replies to my review in the cover letter (the PR description). He edits it after every round of review.

But, when the cover letter is sent as email, that quoting gets messed up and the result looks like [this](https://lore.kernel.org/git/pull.436.v5.git.1574627876.gitgitgadget@gmail.com/).

The first line of a multi-line quoted text is prefixed with '> ', but the rest of the lines aren't. This makes it very difficult to quickly differentiate between the quoted text and reply text.

**Comment by webstech** [comment]
GitHub appears to keep the PR comments in markdown [format.](https://api.github.com/repos/gitgitgadget/git/pulls/436).  GitGitGadget converts the markdown to html and then back to text.  The multi line quotes in markdown become blockquotes in html, with multiple quoted lines converted to a single line.  The conversion from html back to text does not seem to be wrapping as expected and maintaining the quoting.

Some initial feedback while research continues.

**Comment by webstech** [comment]
The fromString function used to convert from html to text does _lazy_ quoting where the `> ` is only on the start of the quoted block.  This relies on mail formatting to support or not support it.

@dscho I have a local override of the blockquote handler if it is of interest.  The [change](https://github.com/gitgitgadget/gitgitgadget/compare/master...webstech:block-quote) does not produce an exact match due to multiple conversions but is pretty close (blank lines between quote levels).  Tested with some actual comments obtained from the api at https://api.github.com/repos/gitgitgadget/gitgitgadget/issues/_nnn_/comments.  Quoting may not happen a lot in a PR (in fact, it is gone from the thread that prompted this issue) but it could be handled better.

**Comment by dscho** [comment]
>  I have a local override of the blockquote handler if it is of interest.

@webstech yes!

> The [change](https://github.com/gitgitgadget/gitgitgadget/compare/master...webstech:block-quote) does not produce an exact match due to multiple conversions but is pretty close (blank lines between quote levels).

I am always a fan of "better than before" 😄

**Comment by dscho** [comment]
```typescript
            const quoteOptions = options;
            quoteOptions.wordwrap -= 2;
```

I think this does not do what you think it does... because `quoteOptions = options` does not do a deep copy, it only copies the reference. See for example:

```sh
$ node -e 'var x = {a: 1, b: 2}; var y = x; y.b = 4; console.log(x); console.log(y);'
{ a: 1, b: 4 }
{ a: 1, b: 4 }
```

I think you need to use `Object.assign({}, options)`:

```sh
$ node -e 'var x = {a: 1, b: 2}; var y = Object.assign({}, x); y.b = 4; console.log(x); console.log(y);'
{ a: 1, b: 2 }
{ a: 1, b: 4 }
```

Note: `Object.assign()` does _not_ do a deep copy, only the first level is copied, any attribute in `options` whose value is an object will _still_ refer to the very same object as before. But in this instance, that's okay, because we only want to override the `wordwrap` attribute and keep all others identical anyway.

About this:

```typescript
return block.replace(/^(>)/, ">$1") // add to quote
                .replace(/^([^>])/, "> $1")     // new quote
                .replace(/\n(>)/g, "\n>$1")
                .replace(/\n([^>\r\n])/g, "\n> $1");
```

I wonder whether we can streamline this a little. All we need here is to prefix every line with "> ", right? So maybe `.replace(/(^|\n)/g, "$&> ")` is the solution? Example:

```sh
$ node -e 'var x = `hello\n> world\n>> or\n>something\n else\n`; console.log(`${x}\n---\n${x.replace(/(^|\n)/g, "$&> ")}`);'
hello
> world
>> or
>something
 else

---
> hello
> > world
> >> or
> >something
>  else
>
```

If you do not want a "> " to be appended after the trailing newline, a negative look-ahead (`(?!$)`: match only if not followed by the end of the string) would be the thing:

```sh
$ node -e 'var x = `hello\n> world\n>> or\n>something\n else\n`; console.log(`${x}\n---\n${x.replace(/(^|\n)(?!$)/g, "$&> ")}`);'
hello
> world
>> or
>something
 else

---
> hello
> > world
> >> or
> >something
>  else

```

**Comment by webstech** [comment]
> I think this does not do what you think it does... because quoteOptions = options does not do a deep copy, it only copies the reference.

Sorry my mistake.  So, how do you feel about wrapping `options.wordwrap' decrement/increment around the `fn()` call?  options is not immutable at this time and it would be cheaper than making a deep copy.  More old school style so thought I would check.

Looking at the replace.  I was not adding the space for embedded blocks (`>> blah` vs `> > blah`).  Will look at at least getting down to two.

**Comment by webstech** [comment]
Pushed up a change.  @dscho I went old school so if you don't like that I will change it before opening the pull request.  There are two replaces - one for the first level and one for the others.

**Comment by dscho** [comment]
> I went old school

You mean non-multi-threadable? ;-)

Seeing as the `-= 2 ... += 2` stanza uses the same number of lines as `Object.assign(...); ... -= 2`, I would prefer the latter.

In particular since we can then make the code safer by requiring a minimum number of columns:

```typescript
const indentedOptions = Object.assign({}, options);
// decrease word wrap, but only to a minimum of 20 columns/line
indentedOptions.wordwrap = Math.max(20, options.wordwrap - 2);
```

> There are two replaces - one for the first level and one for the others.

Makes sense, otherwise you end up with ugly `> > > >`. It is much nicer to read `>>>>` (and it saves half the columns).

About this:

```typescript
            return block.replace(/^>/mg, ">>") // add to quote
                .replace(/^(?!>|$)/mg, "> ");  // new quote
```

I see that the text

```
> Starting a block quote.
>
>> Previously quoted ipsum loren.

> Back to the base quote.
```

is rendered as

```
> Starting a block quote.

>> Previously quoted ipsum loren.

> Back to the base quote.
```

i.e. the quoted empty line is no longer quoted. I think this is because the `(?!>|$)` non-capturing group (or "negative look-ahead") tries to match _neither_ `>` nor the end of the line.

I think, however, that we will want to quote empty lines (without a trailing space), so I'd like to insert a `>` into those. Let's put that to a test:

```sh
$ node -e 'let x = `> a\n>\n\n>> c\n`; console.log(x.replace(/^>/mg, ">>").repla
ce(/^(?!>|$)/mg, "> ").replace(/(^|\n)(\n)/g, "$1>$2"));'
>> a
>>
>
>>> c


```

Note that I do _not_ suggest `.replace(/^$/mg, ">")`, but `.replace(/(^|\n)(\n)/g, "$1>$2")`: the latter avoids appending a `>` after the trailing newline.

**Comment by webstech** [comment]
> You mean non-multi-threadable? ;-)

That is funny .

> ```ts
> const indentedOptions = Object.assign({}, options);
> // decrease word wrap, but only to a minimum of 20 columns/line
> indentedOptions.wordwrap = Math.max(20, options.wordwrap - 2);
> ```

Sometimes it is just too ugly.
```ts
            const indentOptions = Object.assign({}, options,
                //  tslint:disable-next-line:align
                {wordwrap: `${Math.max(20, options.wordwrap - 2)}`});
```

> Note that I do _not_ suggest `.replace(/^$/mg, ">")`, but `.replace(/(^|\n)(\n)/g, "$1>$2")`: the latter avoids appending a `>` after the trailing newline.

Unfortunately, `marked()` generally does `<blockquote><p>ipsem</p></blockquote>` so there is really an extra empty line at the end for the paragraph separator.  As a result, the output ends up as:
```
>> which did not change the purpose of --work-dir.
>>
> From my analysis, it looks as if --work-dir does not actually do
> anything. Or do you find any effect of passing that option to
> misc-helper?
>
Argh. I managed to forget about
```
The updated test also shows this.  Still better than what was happening before.

I will open a pull request to make this more visible in the future.

**Comment by dscho** [comment]
Closed via #168.
